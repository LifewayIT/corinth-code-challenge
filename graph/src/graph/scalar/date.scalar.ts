import {GraphQLError, GraphQLScalarType, Kind} from 'graphql';
import {isDate, isString} from 'lodash';
import {DateTime} from 'luxon';

const _formatStr = 'YYYY-MM-DD';

export const dateScalar = new GraphQLScalarType({

    name: 'Date',

    serialize: (value: Date | string) => {

        if (value === null || typeof value === 'undefined') {
            return null;
        }

        if (isDate(value)) {
            return DateTime.fromJSDate(value).toFormat(_formatStr);
        } else if (isString(value)) {
            return DateTime.fromISO(value).toFormat(_formatStr);
        } else {
            throw new TypeError('Field error: value is an invalid Date');
        }

    },

    parseValue: (value) => {

        const date = DateTime.fromISO(value);
        if (!date.isValid) {
            throw new TypeError('Field error: value is an invalid Date');
        }

        return date.toJSDate();

    },

    parseLiteral: (literal) => {

        if (literal.kind !== Kind.STRING) {
            throw new GraphQLError(`Query error: Can only parse strings to dates but got a: ${literal.kind}`,
                [literal]);
        }

        const result = DateTime.fromISO(literal.value);
        if (!result.isValid) {
            throw new GraphQLError('Query error: Invalid date', [literal]);
        }

        return result.toJSDate();

    }

});
