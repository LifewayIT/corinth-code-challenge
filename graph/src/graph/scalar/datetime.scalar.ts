import {GraphQLError, GraphQLScalarType, Kind} from 'graphql';
import {isDate, isString} from 'lodash';
import {DateTime} from 'luxon';

const _formatStr = 'YYYY-MM-DDTHH:mm:ssZ';

export const dateTimeScalar = new GraphQLScalarType({

    name: 'DateTime',

    serialize: (value: Date | string) => {

        if (value === null || typeof value === 'undefined') {
            return null;
        }

        if (isDate(value)) {
            return DateTime.fromJSDate(value).toFormat(_formatStr);
        } else if (isString(value)) {
            return DateTime.fromISO(value).toFormat(_formatStr);
        } else {
            throw new TypeError('Field error: value is an invalid date/time');
        }

    },

    parseValue: (value) => {

        const date = DateTime.fromISO(value);
        if (!date.isValid) {
            throw new TypeError('Field error: value is an invalid date/time');
        }

        return date.toJSDate();

    },

    parseLiteral: (literal) => {

        if (literal.kind !== Kind.STRING) {
            throw new GraphQLError(`Query error: Can only parse strings to dates but got a: ${literal.kind}`,
                [literal]);
        }

        const result = DateTime.fromISO(literal.value);
        if (!result.isValid) {
            throw new GraphQLError('Query error: Invalid date/time', [literal]);
        }

        return result.toJSDate();

    }

});
