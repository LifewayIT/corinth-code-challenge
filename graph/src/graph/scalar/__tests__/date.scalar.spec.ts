import {Kind} from 'graphql';
import {DateTime} from 'luxon';
import {dateScalar} from '../date.scalar';

describe('dateScalar', () => {

    it('should serialize null/undefined value', () => {

        expect(dateScalar.serialize(null)).toEqual(null);
        expect(dateScalar.serialize(undefined)).toEqual(null);

    });

    it('should serialize Date value', () => {

        const result = '2019-02-11';
        const input = '2019-02-11';

        const date = DateTime.fromISO(input).setZone('America/New_York').toJSDate();
        const str = input;
        const mt = DateTime.fromISO(input);
        const num = 123445;

        expect(dateScalar.serialize(date)).toEqual(result);
        expect(dateScalar.serialize(str)).toEqual(result);
        expect(dateScalar.serialize(mt)).toEqual(result);
        expect(() =>
            dateScalar.serialize(num) as DateTime
        ).toThrowErrorMatchingSnapshot();

    });

    it('should parse DateTime value', () => {

        const input = '2019-01-11';
        expect(dateScalar.parseValue(input)).toMatchSnapshot();

    });

    it('should throw error when parsing DateTime value', () => {

        const input = 'Some weird string';

        expect(() =>
            dateScalar.parseValue(input) as DateTime
        ).toThrowErrorMatchingSnapshot();

    });

    it('should parse literal', () => {

        const input = '2019-01-11';
        expect(dateScalar.parseLiteral({
            kind: Kind.STRING,
            value: input
        }, {})).toMatchSnapshot();

    });

    it('should throw error when parsing non-string literal', () => {

        expect(() =>
            dateScalar.parseLiteral({
                kind: Kind.BOOLEAN,
                value: true
            }, {}) as DateTime
        ).toThrowErrorMatchingSnapshot();

    });

    it('should throw error when parsing literal with bad date', () => {

        const input = 'Some weird string';
        expect(() =>
            dateScalar.parseLiteral({
                kind: Kind.STRING,
                value: input
            }, {}) as DateTime
        ).toThrowErrorMatchingSnapshot();

    });

});
