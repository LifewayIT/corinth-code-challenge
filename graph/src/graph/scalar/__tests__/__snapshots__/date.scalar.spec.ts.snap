// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dateScalar should parse DateTime value 1`] = `2019-01-11T05:00:00.000Z`;

exports[`dateScalar should parse literal 1`] = `2019-01-11T05:00:00.000Z`;

exports[`dateScalar should serialize Date value 1`] = `"Field error: value is an invalid Date"`;

exports[`dateScalar should throw error when parsing DateTime value 1`] = `"Field error: value is an invalid Date"`;

exports[`dateScalar should throw error when parsing literal with bad date 1`] = `"Query error: Invalid date"`;

exports[`dateScalar should throw error when parsing non-string literal 1`] = `"Query error: Can only parse strings to dates but got a: BooleanValue"`;
